server/handlers.go

```diff
@@ -164,10 +164,10 @@ type discovery struct {
 func (s *Server) discoveryHandler() (http.HandlerFunc, error) {
 	d := discovery{
 		Issuer:      s.issuerURL.String(),
-		Auth:        s.absURL("/auth"),
-		Token:       s.absURL("/token"),
-		Keys:        s.absURL("/keys"),
-		UserInfo:    s.absURL("/userinfo"),
+		Auth:        s.absURL(s.publicURL, "/auth"),
+		Token:       s.absURL(s.issuerURL, "/token"),
+		Keys:        s.absURL(s.issuerURL, "/keys"),
+		UserInfo:    s.absURL(s.issuerURL, "/userinfo"),
```

Comment:
Isn't this quite arbitrary? The userinfo endpoint can be called by public clients as well.

Seems to me that the entire feature would be better handled on client side, and not in Dex.

Reply:
Not sure what you mean, /auth is the only thing that needs to be "public" because that's what is used when bouncing users to the underlying backend. The rest are using by authentication consumers, which in many cases are entirely internal.

Reply:
If there is a way to do this client-side without violating the spec, I'm all ears.

Reply:
E.g. a single-page app (using PKCE or Implicit Flow) may call the userinfo endpoint from the browser.
I was assuming that "public URL" = called from browser, "internal URL" = called from backend.

> If there is a way to do this client-side without violating the spec

Not that I'm aware of... Rewriting the URL was just a pragmatic solution that worked for us.

What could also work, if public and internal URLs use the same scheme (HTTP / HTTPS): Add an `/etc/hosts` entry for the backend that should use the internal endpoint, like `<public_dex_hostname> <internal_dex_ip>`.

Reply:
Sure but that SPA would be a client so you wouldn’t enable this mode. This is specifically for the case of all clients are internal but the IdP isn’t (google in my case).